### ISSUE TITLE
For-loop is mis-optimized into while (true)

### ISSUE THREAD

**Issue by HansKristian-ARM** [KhronosGroup/SPIRV-Tools#1184]
Commit: 716718a5e969f6b4e73cbc864db59a754a83aab3

Input:
```
; SPIR-V
; Version: 1.0
; Generator: Khronos Glslang Reference Front End; 2
; Bound: 93
; Schema: 0
               OpCapability Shader
          %1 = OpExtInstImport "GLSL.std.450"
               OpMemoryModel Logical GLSL450
               OpEntryPoint Vertex %main "main" %_ %aVertex %vColor %aNormal
               OpSource ESSL 310
               OpName %main "main"
               OpName %gl_PerVertex "gl_PerVertex"
               OpMemberName %gl_PerVertex 0 "gl_Position"
               OpMemberName %gl_PerVertex 1 "gl_PointSize"
               OpName %_ ""
               OpName %Light "Light"
               OpMemberName %Light 0 "Position"
               OpMemberName %Light 1 "Radius"
               OpMemberName %Light 2 "Color"
               OpName %UBO "UBO"
               OpMemberName %UBO 0 "uMVP"
               OpMemberName %UBO 1 "lights"
               OpName %__0 ""
               OpName %aVertex "aVertex"
               OpName %vColor "vColor"
               OpName %i "i"
               OpName %Light_0 "Light"
               OpMemberName %Light_0 0 "Position"
               OpMemberName %Light_0 1 "Radius"
               OpMemberName %Light_0 2 "Color"
               OpName %light "light"
               OpName %L "L"
               OpName %aNormal "aNormal"
               OpMemberDecorate %gl_PerVertex 0 BuiltIn Position
               OpMemberDecorate %gl_PerVertex 1 BuiltIn PointSize
               OpDecorate %gl_PerVertex Block
               OpMemberDecorate %Light 0 Offset 0
               OpMemberDecorate %Light 1 Offset 12
               OpMemberDecorate %Light 2 Offset 16
               OpDecorate %_arr_Light_uint_4 ArrayStride 32
               OpMemberDecorate %UBO 0 ColMajor
               OpMemberDecorate %UBO 0 Offset 0
               OpMemberDecorate %UBO 0 MatrixStride 16
               OpMemberDecorate %UBO 1 Offset 64
               OpDecorate %UBO Block
               OpDecorate %__0 DescriptorSet 0
               OpDecorate %aVertex Location 0
               OpDecorate %vColor Location 0
               OpDecorate %aNormal Location 1
       %void = OpTypeVoid
          %3 = OpTypeFunction %void
      %float = OpTypeFloat 32
    %v4float = OpTypeVector %float 4
%gl_PerVertex = OpTypeStruct %v4float %float
%_ptr_Output_gl_PerVertex = OpTypePointer Output %gl_PerVertex
          %_ = OpVariable %_ptr_Output_gl_PerVertex Output
        %int = OpTypeInt 32 1
      %int_0 = OpConstant %int 0
%mat4v4float = OpTypeMatrix %v4float 4
    %v3float = OpTypeVector %float 3
      %Light = OpTypeStruct %v3float %float %v4float
       %uint = OpTypeInt 32 0
     %uint_4 = OpConstant %uint 4
%_arr_Light_uint_4 = OpTypeArray %Light %uint_4
        %UBO = OpTypeStruct %mat4v4float %_arr_Light_uint_4
%_ptr_Uniform_UBO = OpTypePointer Uniform %UBO
        %__0 = OpVariable %_ptr_Uniform_UBO Uniform
%_ptr_Uniform_mat4v4float = OpTypePointer Uniform %mat4v4float
%_ptr_Input_v4float = OpTypePointer Input %v4float
    %aVertex = OpVariable %_ptr_Input_v4float Input
%_ptr_Output_v4float = OpTypePointer Output %v4float
     %vColor = OpVariable %_ptr_Output_v4float Output
    %float_0 = OpConstant %float 0
         %33 = OpConstantComposite %v4float %float_0 %float_0 %float_0 %float_0
%_ptr_Function_int = OpTypePointer Function %int
      %int_4 = OpConstant %int 4
       %bool = OpTypeBool
    %Light_0 = OpTypeStruct %v3float %float %v4float
%_ptr_Function_Light_0 = OpTypePointer Function %Light_0
      %int_1 = OpConstant %int 1
%_ptr_Uniform_Light = OpTypePointer Uniform %Light
%_ptr_Function_v3float = OpTypePointer Function %v3float
%_ptr_Function_float = OpTypePointer Function %float
      %int_2 = OpConstant %int 2
%_ptr_Function_v4float = OpTypePointer Function %v4float
%_ptr_Input_v3float = OpTypePointer Input %v3float
    %aNormal = OpVariable %_ptr_Input_v3float Input
    %float_1 = OpConstant %float 1
%_ptr_Uniform_v4float = OpTypePointer Uniform %v4float
       %main = OpFunction %void None %3
          %5 = OpLabel
          %i = OpVariable %_ptr_Function_int Function
      %light = OpVariable %_ptr_Function_Light_0 Function
          %L = OpVariable %_ptr_Function_v3float Function
         %23 = OpAccessChain %_ptr_Uniform_mat4v4float %__0 %int_0
         %24 = OpLoad %mat4v4float %23
         %27 = OpLoad %v4float %aVertex
         %28 = OpMatrixTimesVector %v4float %24 %27
         %30 = OpAccessChain %_ptr_Output_v4float %_ %int_0
               OpStore %30 %28
               OpStore %vColor %33
               OpStore %i %int_0
               OpBranch %36
         %36 = OpLabel
               OpLoopMerge %38 %39 None
               OpBranch %40
         %40 = OpLabel
         %41 = OpLoad %int %i
         %44 = OpSLessThan %bool %41 %int_4
               OpBranchConditional %44 %37 %38
         %37 = OpLabel
         %49 = OpLoad %int %i
         %51 = OpAccessChain %_ptr_Uniform_Light %__0 %int_1 %49
         %52 = OpLoad %Light %51
         %53 = OpCompositeExtract %v3float %52 0
         %55 = OpAccessChain %_ptr_Function_v3float %light %int_0
               OpStore %55 %53
         %56 = OpCompositeExtract %float %52 1
         %58 = OpAccessChain %_ptr_Function_float %light %int_1
               OpStore %58 %56
         %59 = OpCompositeExtract %v4float %52 2
         %62 = OpAccessChain %_ptr_Function_v4float %light %int_2
               OpStore %62 %59
         %64 = OpLoad %v4float %aVertex
         %65 = OpVectorShuffle %v3float %64 %64 0 1 2
         %66 = OpAccessChain %_ptr_Function_v3float %light %int_0
         %67 = OpLoad %v3float %66
         %68 = OpFSub %v3float %65 %67
               OpStore %L %68
         %71 = OpLoad %v3float %aNormal
         %72 = OpLoad %v3float %L
         %73 = OpExtInst %v3float %1 Normalize %72
         %74 = OpDot %float %71 %73
         %76 = OpLoad %v3float %L
         %77 = OpExtInst %float %1 Length %76
         %78 = OpAccessChain %_ptr_Function_float %light %int_1
         %79 = OpLoad %float %78
         %80 = OpFDiv %float %77 %79
         %81 = OpFSub %float %float_1 %80
         %82 = OpExtInst %float %1 FClamp %81 %float_0 %float_1
         %83 = OpLoad %int %i
         %85 = OpAccessChain %_ptr_Uniform_v4float %__0 %int_1 %83 %int_2
         %86 = OpLoad %v4float %85
         %87 = OpVectorTimesScalar %v4float %86 %82
         %88 = OpVectorTimesScalar %v4float %87 %74
         %89 = OpLoad %v4float %vColor
         %90 = OpFAdd %v4float %89 %88
               OpStore %vColor %90
               OpBranch %39
         %39 = OpLabel
         %91 = OpLoad %int %i
         %92 = OpIAdd %int %91 %int_1
               OpStore %i %92
               OpBranch %36
         %38 = OpLabel
               OpReturn
               OpFunctionEnd
```

After:
```
; SPIR-V
; Version: 1.0
; Generator: Khronos Glslang Reference Front End; 2
; Bound: 98
; Schema: 0
               OpCapability Shader
          %1 = OpExtInstImport "GLSL.std.450"
               OpMemoryModel Logical GLSL450
               OpEntryPoint Vertex %main "main" %_ %aVertex %vColor %aNormal
               OpSource ESSL 310
               OpName %main "main"
               OpName %gl_PerVertex "gl_PerVertex"
               OpMemberName %gl_PerVertex 0 "gl_Position"
               OpMemberName %gl_PerVertex 1 "gl_PointSize"
               OpName %_ ""
               OpName %Light "Light"
               OpMemberName %Light 0 "Position"
               OpMemberName %Light 1 "Radius"
               OpMemberName %Light 2 "Color"
               OpName %UBO "UBO"
               OpMemberName %UBO 0 "uMVP"
               OpMemberName %UBO 1 "lights"
               OpName %__0 ""
               OpName %aVertex "aVertex"
               OpName %vColor "vColor"
               OpName %Light_0 "Light"
               OpMemberName %Light_0 0 "Position"
               OpMemberName %Light_0 1 "Radius"
               OpMemberName %Light_0 2 "Color"
               OpName %aNormal "aNormal"
               OpMemberDecorate %gl_PerVertex 0 BuiltIn Position
               OpMemberDecorate %gl_PerVertex 1 BuiltIn PointSize
               OpDecorate %gl_PerVertex Block
               OpMemberDecorate %Light 0 Offset 0
               OpMemberDecorate %Light 1 Offset 12
               OpMemberDecorate %Light 2 Offset 16
               OpDecorate %_arr_Light_uint_4 ArrayStride 32
               OpMemberDecorate %UBO 0 ColMajor
               OpMemberDecorate %UBO 0 Offset 0
               OpMemberDecorate %UBO 0 MatrixStride 16
               OpMemberDecorate %UBO 1 Offset 64
               OpDecorate %UBO Block
               OpDecorate %__0 DescriptorSet 0
               OpDecorate %aVertex Location 0
               OpDecorate %vColor Location 0
               OpDecorate %aNormal Location 1
       %void = OpTypeVoid
          %3 = OpTypeFunction %void
      %float = OpTypeFloat 32
    %v4float = OpTypeVector %float 4
%gl_PerVertex = OpTypeStruct %v4float %float
%_ptr_Output_gl_PerVertex = OpTypePointer Output %gl_PerVertex
          %_ = OpVariable %_ptr_Output_gl_PerVertex Output
        %int = OpTypeInt 32 1
      %int_0 = OpConstant %int 0
%mat4v4float = OpTypeMatrix %v4float 4
    %v3float = OpTypeVector %float 3
      %Light = OpTypeStruct %v3float %float %v4float
       %uint = OpTypeInt 32 0
     %uint_4 = OpConstant %uint 4
%_arr_Light_uint_4 = OpTypeArray %Light %uint_4
        %UBO = OpTypeStruct %mat4v4float %_arr_Light_uint_4
%_ptr_Uniform_UBO = OpTypePointer Uniform %UBO
        %__0 = OpVariable %_ptr_Uniform_UBO Uniform
%_ptr_Uniform_mat4v4float = OpTypePointer Uniform %mat4v4float
%_ptr_Input_v4float = OpTypePointer Input %v4float
    %aVertex = OpVariable %_ptr_Input_v4float Input
%_ptr_Output_v4float = OpTypePointer Output %v4float
     %vColor = OpVariable %_ptr_Output_v4float Output
    %float_0 = OpConstant %float 0
         %33 = OpConstantComposite %v4float %float_0 %float_0 %float_0 %float_0
%_ptr_Function_int = OpTypePointer Function %int
      %int_4 = OpConstant %int 4
       %bool = OpTypeBool
    %Light_0 = OpTypeStruct %v3float %float %v4float
%_ptr_Function_Light_0 = OpTypePointer Function %Light_0
      %int_1 = OpConstant %int 1
%_ptr_Uniform_Light = OpTypePointer Uniform %Light
%_ptr_Function_v3float = OpTypePointer Function %v3float
%_ptr_Function_float = OpTypePointer Function %float
      %int_2 = OpConstant %int 2
%_ptr_Function_v4float = OpTypePointer Function %v4float
%_ptr_Input_v3float = OpTypePointer Input %v3float
    %aNormal = OpVariable %_ptr_Input_v3float Input
    %float_1 = OpConstant %float 1
%_ptr_Uniform_v4float = OpTypePointer Uniform %v4float
       %true = OpConstantTrue %bool
       %main = OpFunction %void None %3
          %5 = OpLabel
         %23 = OpAccessChain %_ptr_Uniform_mat4v4float %__0 %int_0
         %24 = OpLoad %mat4v4float %23
         %27 = OpLoad %v4float %aVertex
         %28 = OpMatrixTimesVector %v4float %24 %27
         %30 = OpAccessChain %_ptr_Output_v4float %_ %int_0
               OpStore %30 %28
               OpStore %vColor %33
               OpBranch %36
         %36 = OpLabel
               OpLoopMerge %38 %39 None
               OpBranch %40
         %40 = OpLabel
               OpBranchConditional %true %37 %38
         %37 = OpLabel
         %51 = OpAccessChain %_ptr_Uniform_Light %__0 %int_1 %int_0
         %52 = OpLoad %Light %51
         %53 = OpCompositeExtract %v3float %52 0
         %56 = OpCompositeExtract %float %52 1
         %65 = OpVectorShuffle %v3float %27 %27 0 1 2
         %68 = OpFSub %v3float %65 %53
         %71 = OpLoad %v3float %aNormal
         %73 = OpExtInst %v3float %1 Normalize %68
         %74 = OpDot %float %71 %73
         %77 = OpExtInst %float %1 Length %68
         %80 = OpFDiv %float %77 %56
         %81 = OpFSub %float %float_1 %80
         %82 = OpExtInst %float %1 FClamp %81 %float_0 %float_1
         %85 = OpAccessChain %_ptr_Uniform_v4float %__0 %int_1 %int_0 %int_2
         %86 = OpLoad %v4float %85
         %87 = OpVectorTimesScalar %v4float %86 %82
         %88 = OpVectorTimesScalar %v4float %87 %74
         %89 = OpLoad %v4float %vColor
         %90 = OpFAdd %v4float %89 %88
               OpStore %vColor %90
               OpBranch %39
         %39 = OpLabel
               OpBranch %36
         %38 = OpLabel
               OpReturn
               OpFunctionEnd
```

**Comment by greg-lunarg** [comment]
Not sure just what recipe you are using, but I believe your issue is resolved by commit e5560d64de40bcc52b38ee44fd8a0bbc42f64399. Can you confirm?

**Comment by dneto0** [comment]
I agree with @greg-lunarg that this was fixed in --ccp.
I reran the input against spirv-opt -O and it keeps the loop induction variable intact.

Closing.
